server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java

```diff
@@ -1484,10 +1487,18 @@ private void cleanupOldShardGens(
                 toDelete.add(containerPath + shardGeneration);
             }
         }
-        try {
-            deleteFromContainer(blobContainer(), toDelete.iterator());
-        } catch (Exception e) {
-            logger.warn("Failed to clean up old shard generation blobs", e);
+        if (toDelete.isEmpty() == false) {
+            threadPool.executor(ThreadPool.Names.SNAPSHOT).execute(() -> {
```

Comment:
Since we compute the list of files to delete based on the old/new repository data, would it be possible for two delete operations to try and delete an intersection of files? I guess that will be fine, the only minor issue I can see is that we might end up having more queued tasks in the `SNAPSHOT` thread pool, but the risk should be minimal?

Reply:
> would it be possible for two delete operations to try and delete an intersection of files?

No this should not happen since we compute the files to delete from the repo data diff. So if a file A goes out of scope when moving from generation 1 to 2, then it will never show up in generation 2 and thus won't be in the diff from 2 to 3.

> I guess that will be fine, the only minor issue I can see is that we might end up having more queued tasks in the SNAPSHOT thread pool, but the risk should be minimal?

This is actually something that happens in the many-shards snapshot benchmark. Assuming you have back to back snapshot finalizations, it's really a function of how fast cluster state updates are vs how fast those deletes are. But it's not really an issue IMO, we're still only talking about O(10s) tasks here in the worst cases (GCP + 50k indices).